validator/client/beacon-api/activation.go

```diff
@@ -0,0 +1,158 @@
+	"google.golang.org/grpc"
+)
+
+func (c beaconApiValidatorClient) waitForActivation(ctx context.Context, in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	// Represents the set of keys in `in`
+	targetPubKeys := make(map[string]struct{})
+
+	// Represents the set of keys actually returned by the beacon node
+	retrievedPubKeys := make(map[string]struct{})
+
+	// Contains all keys in targetPubKeys but not in retrievedPubKeys
+	missingPubKeys := make(map[string]struct{})
+
+	statuses := []*ethpb.ValidatorActivationResponse_Status{}
+
+	for _, publicKey := range in.PublicKeys {
+		stringPubKey := "0x" + hex.EncodeToString(publicKey)
+		targetPubKeys[stringPubKey] = struct{}{}
+	}
+
+	params := url.Values{}
+
+	for _, publicKey := range in.PublicKeys {
+		params.Add("id", fmt.Sprintf("0x%s", hex.EncodeToString(publicKey)))
+	}
+
+	resp, err := c.httpClient.Get(c.url + "/eth/v1/beacon/states/head/validators?" + params.Encode())
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		errorJson := apimiddleware.DefaultErrorJson{}
+		err = json.NewDecoder(resp.Body).Decode(&errorJson)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, errors.Errorf("error %d: %s", errorJson.Code, errorJson.Message)
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+
+	var res ValidatorsHttpResponse
+	if err := json.Unmarshal(body, &res); err != nil {
+		return nil, err
+	}
+
+	for _, data := range res.Data {
+		pubkey, err := hex.DecodeString(data.Validator.Pubkey[2:])
+
+		if err != nil {
+			return nil, err
+		}
+
+		retrievedPubKeys[data.Validator.Pubkey] = struct{}{}
+
+		index, err := strconv.ParseUint(data.Index, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		validatorStatus, ok := BeaconAPITogRPCValidatorStatus[data.Status]
+		if !ok {
+			return nil, errors.New("Unknown validator status: " + data.Status)
+		}
+
+		statuses = append(statuses, &ethpb.ValidatorActivationResponse_Status{
+			PublicKey: pubkey,
+			Index:     types.ValidatorIndex(index),
+			Status:    &ethpb.ValidatorStatusResponse{Status: validatorStatus},
+		})
+	}
+
+	for targetPubKey := range targetPubKeys {
+		if _, ok := retrievedPubKeys[targetPubKey]; !ok {
+			missingPubKeys[targetPubKey] = struct{}{}
+		}
+	}
+
+	for missingPubKey := range missingPubKeys {
+		pubkeyBytes, err := hex.DecodeString(missingPubKey[2:])
+		if err != nil {
+			return nil, err
+		}
+
+		statuses = append(statuses, &ethpb.ValidatorActivationResponse_Status{
+			PublicKey: pubkeyBytes,
+			Index:     types.ValidatorIndex(^uint64(0)),
+			Status:    &ethpb.ValidatorStatusResponse{Status: ethpb.ValidatorStatus_UNKNOWN_STATUS},
+		})
+	}
+
+	return &beaconNodeValidatorWaitForActivationClient{
+		ValidatorActivationResponse: ethpb.ValidatorActivationResponse{
+			Statuses: statuses,
+		},
+	}, nil
+}
+
+type ValidatorsHttpResponse struct {
```

Comment:
We should be able to use `StateValidatorsResponseJson` from `beacon-chain/rpc/apimiddleware/structs.go`

Reply:
Perfect! It's exactly what I was looking for!